File: subprojects/studio-plugin/src/main/java/org/gradle/profiler/studio/plugin/GradleProfilerStartupActivity.java
package org.gradle.profiler.studio.plugin;

import com.android.tools.idea.gradle.project.GradleProjectInfo;
import com.intellij.ide.impl.TrustedProjects;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.externalSystem.service.notification.ExternalSystemProgressNotificationManager;
import com.intellij.openapi.externalSystem.settings.ExternalSystemSettingsListenerAdapter;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import org.gradle.profiler.client.protocol.Client;
import org.gradle.profiler.client.protocol.messages.StudioRequest;
import org.gradle.profiler.studio.plugin.client.GradleProfilerClient;
import org.gradle.profiler.studio.plugin.system.AndroidStudioSystemHelper;
import org.gradle.profiler.studio.plugin.system.GradleSystemListener;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.plugins.gradle.settings.GradleProjectSettings;
import org.jetbrains.plugins.gradle.settings.GradleSettings;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Collection;

import static org.gradle.profiler.client.protocol.messages.StudioRequest.StudioRequestType.EXIT_IDE;

public class GradleProfilerStartupActivity implements StartupActivity {

    private static final Logger LOG = Logger.getInstance(GradleProfilerStartupActivity.class);

    public static final String PROFILER_PORT_PROPERTY = "gradle.profiler.port";

    @Override
    public void runActivity(@NotNull Project project) {
        LOG.info("Project opened");
        if (System.getProperty(PROFILER_PORT_PROPERTY) != null) {
            // This solves the issue where Android Studio would run the Gradle sync automatically on the first import.
            // Unfortunately it seems we can't always detect it because it happens very late and due to that there might
            // a case where two simultaneous syncs would be run: one from automatic sync trigger and one from our trigger.
            // With this line we disable that automatic sync, but we still trigger our sync later in the code.
            GradleProjectInfo.getInstance(project).setSkipStartupActivity(true);
            // If we don't disable external annotations, Android Studio will download some artifacts
            // to .m2 folder if some project has for example com.fasterxml.jackson.core:jackson-core as a dependency
            disableDownloadOfExternalAnnotations(project);
            // Register system listener already here, so we can catch any failure for syncs that are automatically started
            GradleSystemListener gradleSystemListener = registerGradleSystemListener();
            TrustedProjects.setTrusted(project, true);
            ApplicationManager.getApplication().executeOnPooledThread(() -> {
                StudioRequest lastRequest = listenForSyncRequests(project, gradleSystemListener);
                if (lastRequest.getType() == EXIT_IDE) {
                    AndroidStudioSystemHelper.exit(project);
                }
            });
        }
    }

    private void disableDownloadOfExternalAnnotations(Project project) {
        GradleSettings gradleSettings = GradleSettings.getInstance(project);
        gradleSettings.getLinkedProjectsSettings()
            .forEach(settings -> settings.setResolveExternalAnnotations(false));
        gradleSettings.subscribe(new ExternalSystemSettingsListenerAdapter<>() {
            @Override
            public void onProjectsLinked(@NotNull Collection<GradleProjectSettings> linkedProjectsSettings) {
                linkedProjectsSettings.forEach(settings -> settings.setResolveExternalAnnotations(false));
            }
        }, gradleSettings);
    }

    private GradleSystemListener registerGradleSystemListener() {
        GradleSystemListener gradleSystemListener = new GradleSystemListener();
        ExternalSystemProgressNotificationManager.getInstance().addNotificationListener(gradleSystemListener);
        return gradleSystemListener;
    }

    private StudioRequest listenForSyncRequests(@NotNull Project project, @NotNull GradleSystemListener gradleStartupListener) {
        int port = Integer.getInteger(PROFILER_PORT_PROPERTY);
        try (Client client = new Client(port)) {
            return new GradleProfilerClient(client).listenForSyncRequests(project, gradleStartupListener);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: subprojects/studio-plugin/src/main/kotlin/org/gradle/profiler/studio/plugin/GradleProfilerStartupActivity.kt
package org.gradle.profiler.studio.plugin

import com.intellij.ide.impl.setTrusted
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.externalSystem.service.notification.ExternalSystemProgressNotificationManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.ProjectActivity
import com.intellij.openapi.util.registry.Registry
import com.intellij.openapi.util.registry.RegistryValue
import org.gradle.profiler.client.protocol.Client
import org.gradle.profiler.client.protocol.messages.IdeRequest
import org.gradle.profiler.client.protocol.messages.IdeRequest.IdeRequestType
import org.gradle.profiler.studio.plugin.client.GradleProfilerClient
import org.gradle.profiler.studio.plugin.system.IdeSystemHelper
import org.gradle.profiler.studio.plugin.system.GradleSystemListener
import org.jetbrains.plugins.gradle.settings.GradleProjectSettings
import org.jetbrains.plugins.gradle.settings.GradleSettings
import java.io.File
import java.io.FileInputStream
import java.util.*

class GradleProfilerStartupActivity : ProjectActivity {
    companion object {
        private val LOG = Logger.getInstance(GradleProfilerStartupActivity::class.java)
        const val PROFILER_PORT_PROPERTY: String = "gradle.profiler.port"
    }

    override suspend fun execute(project: Project) {
        LOG.info("Project opened")
        logModifiedRegistryEntries()
        if (System.getProperty(PROFILER_PORT_PROPERTY) == null) return

        configureGradleSettings(project)
        // Register system listener already here, so we can catch any failure for syncs that are automatically started
        val gradleSystemListener = GradleSystemListener()
        ExternalSystemProgressNotificationManager.getInstance().addNotificationListener(gradleSystemListener, project)

        project.setTrusted(true)

        ApplicationManager.getApplication().executeOnPooledThread {
            val lastRequest = listenForSyncRequests(project, gradleSystemListener)
            if (lastRequest.type == IdeRequestType.EXIT_IDE) {
                IdeSystemHelper.exit(project)
            }
        }
    }

    private fun logModifiedRegistryEntries() {
        val studioPropertiesPath = System.getenv("STUDIO_PROPERTIES")
        if (studioPropertiesPath == null || !File(studioPropertiesPath).exists()) {
            return
        }

        val properties = Properties().apply { this.load(FileInputStream(studioPropertiesPath)) }
        val modifiedValues = Registry.getAll()
            .filter { value: RegistryValue -> properties.containsKey(value.key) }
            .map { obj: RegistryValue -> obj.toString() }
            .sorted()
        LOG.info("Modified registry entries: $modifiedValues")
    }

    private fun configureGradleSettings(project: Project) {
        val gradleSettings = GradleSettings.getInstance(project)
        gradleSettings.linkedProjectsSettings.forEach {
            configureLinkedProject(it)
        }
        gradleSettings.subscribe(object : DefaultGradleSettingsListener() {
            override fun onProjectsLinked(linkedProjectsSettings: Collection<GradleProjectSettings>) {
                linkedProjectsSettings.forEach { configureLinkedProject(it) }
            }
        }, gradleSettings)
    }

    private fun configureLinkedProject(settings: GradleProjectSettings) {
        // If we don't disable external annotations, the IDE will download some artifacts
        // to .m2 folder if some project has for example com.fasterxml.jackson.core:jackson-core as a dependency
        settings.isResolveExternalAnnotations = false
        // Set Gradle JVM to JAVA_HOME to avoid JDK resolution dialogs in headless mode
        if (settings.gradleJvm == null || settings.gradleJvm == "#USE_PROJECT_JDK") {
            settings.gradleJvm = "#JAVA_HOME"
            LOG.info("Set Gradle JVM to #JAVA_HOME for ${settings.externalProjectPath}")
        }
    }

    private fun listenForSyncRequests(project: Project, gradleStartupListener: GradleSystemListener): IdeRequest {
        val port = Integer.getInteger(PROFILER_PORT_PROPERTY)
        Client(port).use {
            return GradleProfilerClient(it).listenForSyncRequests(project, gradleStartupListener)
        }
    }
}
